MOAR Slash Commands, Titles, AFKs, ChatMessages, String Replacements, and other unlabeled-bits

broxen opened the pull request:
Many small change in this one. Just trying to move the ball forward.

Additions/modifications proposed in this pull request:
- Attempting to track down unlabeled bits and their purpose. 
- Added slash commands for fly, dazed, jumppack, jumpheight, and backupspd, AFK, plus more.
- JumpHeight might work in tandem with u3 to enable "superjump" powers.
- Added ChatMessage for Request, Friends, Group, Supergroup, and associated Slash commands
- Pulled description and battle_cry out of entity. Fixed them so they update in your ID card properly. yay!
- Added handler for selecting different chat channels from the chat window input box. code 37
- TODO: still need to save things to the database, Titles, Desc, Battle_Cry, MapName, etc.
- TODO: Jumppack and Dazed still need their associated costume parts and special FX (stars above head) respectively.
- TODO: tried to add handler for client message 32, but it doesn't work, maybe I'm missing something?

The following slash commands now work in the client:
```
/f
/req
/sg
/g
/fly
/dazed
/jumppack
/setJumpHeight [float]
/setBackupspd [float]
/afk [msg]
/setTitles [the_flag] [generic] [origin] [special]
/setu1 [int]
/setu2 [int]
/setu3 [int]
/setu4 [int]
/setu5 [int]
/setu6 [int]
```

There are still many unknown bits here, but I have suspicions about these two in SendServerControlState()
```
        bs.StoreBits(1,u2);     // if 1/true entity anims stop, can still move, but camera stays. Slipping on ice?
        bs.StoreBits(1,u3);     // leaping? seems like the anim changes slightly?
```

EDIT: Typos.

broxen commented:
Oh neato. It carries my own comments over.

Well the crashing issues are fixed, but I always encourage testing. Also, please feel free to let me know if there are better ways of achieving my goals, and if you know what any of these unknown bits do.

nemerle commented:
This looks great.

Tomorrow should be the last day of that big at-work thing I'm currently busy with, so the day after I'l look through my resources for all those u1/u2 etc. bits and pieces and merge this

mobbyg commented:
Hey guys, thank you for the work you’ve done on this project. It’s really great reading the emails for the commits. I try and update server as soon as you post changes so I can offer any info I can, as I don’t know C++, but I have a For Dummies book now. 😜

broxen commented:
Great. Thank you both!

When you get a chance to look this over, can you tell me how I can call `toggleAFK()` when a player moves? Seems like it should go in `sendServerControlState()` but I can't access Entity there because it's `Const` (and rightly so)

Intended behavior is that movement (WASD, or mouse) will turn off AFK flag.

broxen commented:
Hope you don't mind me pushing another update to this PR.

- Added more slash commands for testing.
- Also, I realized that I need to set m_afk to false on movement, not Toggle. So if you can point me in the right direction for that, I'd appreciate it.
- Identified and began framework for opcodes 2, 8, 18, 29
- Also fixed opcode 16, which was somehow removed from MapEventFactory.cpp

nemerle commented on Projects/CoX/Common/NetStructures/Character.cpp:
```diff
@@ -114,10 +88,27 @@ void Character::setName(const QString &val )
+
+void Character::setTitles(bool prefix, QString generic, QString origin, QString special)
+{
+    if(!prefix && generic.isEmpty() && origin.isEmpty() && special.isEmpty())
```
consider writing this like ?
```cpp
m_has_titles = prefix || !generic.isEmpty() || !origin.isEmpty() || special.isEmpty();
if(!m_has_titles)
  return;
```

broxen replied:
Ooo. Much cleaner!

nemerle commented on Projects/CoX/Common/NetStructures/Character.cpp:
```diff
@@ -688,3 +679,27 @@ void ClientOptions::init()
+void Character::toggleAFK(const QString &msg)
+{
+    m_afk = !m_afk;
+    m_afk_msg = msg;
```
Do we need to clean up m_afk_msg if client is no longer afk ?

broxen replied:
Testing this: no, it doesn't seem like it matters. But my (admittedly terrible) memory tells me that the official server behavior was actually to save this msg between uses (during the same connection)

I think that because I used to go afk with msg, move (removing afk flag) and then set afk flag again without msg and it would have my old msg left in the system (probably client side memory).

I guess I could achieve this with a simple `if` statement, but is there any reason (memory) to clear this value instead?

nemerle commented on Projects/CoX/Common/NetStructures/Character.h:
```diff
@@ -175,13 +177,36 @@ const   QString &       getClass() const { return m_class_name; }
+        QString             m_character_description;
+        bool                m_afk               = false;
+        QString             m_afk_msg;
+        // Experience Requirements Chart. Same for all classes.
```
We should load it from the data files anyway :)
If everything goes right today, I'll be pushing code to access those kinds of things.

nemerle replied:
Huh, You're right, I'd swear I've seen the exp table somewhere

nemerle replied:
Yup, it's in experience.bin added issues #176

broxen replied:
Seems like it would be, there's no real harm in exposing this info client side. And, you've looked through more of the `bin` files than me, so I'm sure you're right, it's there somewhere.

nemerle commented on Projects/CoX/Servers/MapServer/MapEvents.h:
```diff
@@ -376,6 +378,42 @@ class SetDestination final : public MapLinkEvent
+    void serializefrom(BitStream &bs) override
+    {
+        entity_idx  = bs.GetPackedBits(12);
+        boost_unk1  = bs.GetAvailSize();  // How many bits!?
```
please mark things like this with
```cpp
//TODO:
```
Qtcreator has a very nice plugin that can list allo TODO: / BUG: etc. markers in the project :)

broxen replied:
Oh neato! I'll be more consistent with this then.

nemerle commented on Projects/CoX/Servers/MapServer/MapInstance.cpp:
```diff
@@ -469,27 +481,58 @@ void MapInstance::on_window_state(WindowState * ev){
+            msg.startsWith("f ")    || msg.startsWith("friends ")    ||
+            msg.startsWith("t ")    || msg.startsWith("tell ")       || msg.startsWith("w ")       || msg.startsWith("whisper ")  ||
+            msg.startsWith("p ")    || msg.startsWith("private ");
 }
```
Maybe we could simplify this by using QStringList::contains here ?
```cpp
static const QStringList chat_prefixes = { "l","local","b","broadcast",/* the rest would follow */ };
QString prefix(msg.mid(0,indexOf(' '));
return chat_prefixes.contains(prefix);
```

broxen replied:
Thank you! I was struggling to figure out a better way to do this, and my brain just wasn't going anywhere.

broxen commented:
Updated based upon feedback, and added additional TODO comments and cleaned up the comments overall.

Also, I cleaned up the code a tiny bit.

nemerle commented on Projects/CoX/Servers/MapServer/Events/MapEventTypes.h:
```diff
@@ -5,43 +5,48 @@ class MapEventTypes : public CRUD_EventTypes
+    EVENT_DECL(evInspirationDockMode        ,217)
+    EVENT_DECL(evPowersDockMode             ,218)
+    EVENT_DECL(evUseInspiration             ,229) // Hit F1, F2, F3, F4, F5 or Click on Inspiration in Tray
+    EVENT_DECL(evSetTarget                  ,232) // Click on another entity
```
I'm almost 100% positive that command 32 is related to unqueing a power or inspiration execution:
Take a look at client code:
```cpp
void sendPacket32(Entity *a1)
{
    my_pakSendPackedBits(g_input_pak, 1, 32);
}
void  AbortQueuedPower()
{
    Entity *e = getControlledPlayerEntity();
    sendPacket32(e);
    inspirationUnqueue();
    g_cursor.target_world = 0;
    s_ptsActivated = nullptr;
}
```

broxen replied:
I actually read this a different way, informed by the client behavior during testing.

I think command 32 is still related to targeting, as we already have `EVENT_DECL(evAbortQueuedPower           ,233)` which fires when you hit 'Z' (bound to unqueue powers by default)

But I think `AbortQueuedPower` is a function called when you hit `ESC` and is not related to command 32, but simply calls it's send function.

This would make sense because hitting `ESC` should do all of the following:
1. set target to self 
```cpp
    Entity *e = getControlledPlayerEntity();
    sendPacket32(e);
```
2. unqueue any powers or inspirations you've activated `inspirationUnqueue();`
3. clear any active waypoints/markers on the map (presumably) `s_ptsActivated = nullptr;`

broxen replied:
latest push appears to correct this behavior here, and supports my supposition in that hitting `ESC` triggers both `SetTarget` and `AbortQueuedPower` resolving #174

nemerle commented on Projects/CoX/Servers/MapServer/MapEvents.h:
```diff
@@ -376,6 +378,43 @@ class SetDestination final : public MapLinkEvent
+    {
+        entity_idx  = bs.GetPackedBits(12);
+        boost_unk1  = bs.GetAvailSize();  // TODO: Not all bits were consumed
+    }
```
Packet 29 - activate inspiration slot  ( it does not contain entity_idx)
packedbits(1,29);
packedbits(3,slot_idx);
packedbits(3,row_idx);

nemerle commented on Projects/CoX/Servers/MapServer/MapEvents.h:
```diff
@@ -454,7 +508,52 @@ class ChatReconfigure final : public MapLinkEvent
         m_chat_bottom_flags = bs.GetPackedBits(1);
     }
 };
+class PowersDockMode final : public MapLinkEvent
```
Packet 18 is used when a secondary tray is toggled:
* packedBits(1,18)
* bits(1,toggle_secondary_tray)

broxen replied:
This comment doesn't show as cleared, but is resolved in the latest push.

nemerle commented on Projects/CoX/Servers/MapServer/MapEvents.h:
```diff
@@ -454,7 +508,52 @@ class ChatReconfigure final : public MapLinkEvent
+        // TODO: Not all bits were consumed
+    }
+};
+class SwitchTray final : public MapLinkEvent
```
Packet 8 contains serialized PowerTrayGroup

nemerle commented on Projects/CoX/Common/NetStructures/Character.cpp:
```diff
@@ -443,16 +432,16 @@ void Character::sendChatSettings(BitStream &bs) const
+    storeStringConditional(bs, m_titles[0]);    // Title 1 - generic title (first)
+    storeStringConditional(bs, m_titles[1]);    // Title 2 - origin title (second)
+    storeStringConditional(bs, m_titles[2]);    // Title 3 - yellow title (special)
 }
```
Sadly this is incorrect vs. the client code:
```cpp
void  receiveTitles(DbPacket *pak)
{
    Entity *e = my_getControlledPlayerEntity();
    e->ent_player2->titleThe = my_pakGetBits(pak, 1);
    strncpy(e->ent_player2->Title1, my_pakGetString(pak), 32);
    strncpy(e->ent_player2->Title2, my_pakGetString(pak), 32);
    strncpy(e->ent_player2->Title3, my_pakGetString(pak), 128);
}
```
So if You had to convert it to conditional strings, then we have a problem somewhere before the sendTitles

broxen replied:
Ah, it just clicked for me how `storeStringConditonal` actually works. 

Not sure where I got the idea to use this, but I can confirm that `/setTitles bFlag generic origin special` actually works.

Maybe the client expects a has_titles flag before this group is received?

broxen replied:
Can also confirm that prior to using Conditional, the client would crash sporadically and this was my best fix.

broxen replied:
I pulled this from sendTitles in Entity (which is duplicitous and I have now removed)

I've tried to send bs.StoreString() instead, but it crashes the client =/

Any ideas?

broxen replied:
Also, why `e->ent_player2` instead of player1

nemerle commented on Projects/CoX/Servers/MapServer/Events/EntitiesResponse.cpp:
```diff
@@ -278,33 +290,35 @@ void sendServerControlState(const EntitiesResponse &src,BitStream &bs)
+        bs.StoreBits(1,m_is_stunned);     // is_stunned flag (lacks overhead 'dizzy' FX)
+        bs.StoreBits(1,m_has_jumppack);  // jumpack flag (lacks costume parts)
+
+        bs.StoreBits(1,u2); // if 1/true entity anims stop, can still move, but camera stays. Slipping on ice?
```
this is likely "controls disabled" bit

broxen replied:
That makes sense. Though you can still move?

broxen commented:
I'll try to rebase this and resolve those conflicts in EntitiesResponse.cpp late tonight.

broxen commented:
Note that this `update_part_2` section is completely wrong. Setting update_part_2 flag to true results in wild behavior, including flashing of location and the entity being relocated to `/loc 3.0 0.0 3.0`

Maybe this section has to do with teleporting?
- m_if_teleporting
- set client pos id (maybe a reference to current location?)
- speed? maybe actually pos, since currently it sends you to 3.0 0.0 3.0
- zeroes? maybe flags for FX? maybe a point in space to direct you to for teleport?
- set orientation upon arrival
- no idea on u6?

```
    // Used to force the client to a position/speed/pitch/rotation by server
    bs.StoreBits(1,update_part_2);
    if(update_part_2)
    {
        bs.StorePackedBits(1,u5);            // sets g_client_pos_id_rel default = 0
        storeVector(bs,spd);
        storeVectorConditional(bs,zeroes);      // vector3 -> speed ? likely; was zeroes

        storeFloatConditional(bs,0);         // Pitch not used ?
        storeFloatConditional(bs,ent->inp_state.camera_pyr.y); // Pitch
        storeFloatConditional(bs,0);         // Roll
        bs.StorePackedBits(1,u6);            // sets the lowest bit in CscCommon::flags default = 0
    }
```

broxen commented:
Lots of updates here, but I couldn't figure out the sendTitles() storeStringCondtional() issue.

broxen commented:
@nemerle any ideas on the sendTitles() issue? I can't make that work without storeStringConditional()

Also, is there anything else you need me to do to prepare this for commit? I think it's ready.

broxen commented:
Pulled many functions out of classes and added some utility functions to get Entity object by IDX or Name. Needs testing though, as I'm on a chromebook until the New Year.

nemerle commented on Projects/CoX/Common/GameData/attrib_serializers.cpp:
```diff
@@ -159,10 +163,14 @@ void serialize(Archive & archive, Parse_CharAttribMax & target)
+=======
     archive(cereal::make_nvp("Jumppack",target.m_Jumppack));
+>>>>>>> upstream/master
     archive(cereal::make_nvp("Teleport",target.m_Teleport));
```
This looks like You've commited a file with a merge conflict :)

broxen replied:
Not sure how that happened because I don't see it on my local branch? Also my local fork shows the variable as `m_has_jumppack` following our naming guidelines. Weird.

Anyway, looks like it's resolved :man_shrugging:

broxen commented on Projects/CoX/Servers/GameServer/GameServer.cpp:
```diff
@@ -229,12 +229,36 @@ GameServerData &GameServer::runtimeData()
+    EntityManager ref_em;
+
+    // Iterate through all active entities and return entity by idx
+    for (auto *em : ref_em.m_live_entlist)
```
Not sure how else to poll for entities in GameServer.

broxen commented on Projects/CoX/Servers/GameServer/GameServer.cpp:
```diff
@@ -229,12 +229,36 @@ GameServerData &GameServer::runtimeData()
+    EntityManager ref_em;
+
+    // Iterate through all active entities and return entity by name
+    for (auto *em : ref_em.m_live_entlist)
```
Not sure how else to poll for entities in GameServer.

broxen commented on Projects/CoX/Servers/AdminServer/CharacterDatabase.cpp:
```diff
@@ -298,31 +301,30 @@ bool CharacterDatabase::update( Character *c )
         "specialtitle=:specialtitle, supergroupid=:supergroupid "
         "WHERE id=:id "
     */
-    m_prepared_char_update.bindValue(":id", getDbId(e));
```
Here I need to access the Entity that Character is associated with, because I need to access e->m_db_id;

How else could I do that?

broxen commented on Projects/CoX/Servers/AdminServer/CharacterDatabase.cpp:
```diff
@@ -298,31 +301,30 @@ bool CharacterDatabase::update( Character *c )
     m_prepared_char_update.bindValue(":xppatrol", c->m_experience_patrol);
-    m_prepared_char_update.bindValue(":alignment", e->m_alignment);
+    m_prepared_char_update.bindValue(":alignment", c->m_alignment);
     m_prepared_char_update.bindValue(":posx", e->pos.x);
```
Here we need to access entity again, to save pos values.

broxen commented on Projects/CoX/Servers/MapServer/MapInstance.cpp:
```diff
@@ -751,23 +755,23 @@ void MapInstance::on_console_command(ConsoleCommand * ev)
     }
     else if(lowerContents == "fly") {
-        toggleFly(ent);
+        toggleFly(*ent);
```
Since pulling these functions out of the entity/character classes I have to pass a pointer to the entity/character, but I'm not understanding why. The functions all take references.

broxen commented on Projects/CoX/Servers/AdminServer/CharacterDatabase.cpp:
```diff
@@ -155,8 +158,8 @@ bool CharacterDatabase::fill( Character *c)
 
-    c->setDbId(m_prepared_char_select.value("id").toUInt());
-    c->setLevel((uint8_t)m_prepared_char_select.value("char_level").toUInt());
+    //setDbId(c,m_prepared_char_select.value("id").toUInt());
```
Here I need Entity again to ensure that the m_db_id value is set from the database.

broxen commented on Projects/CoX/Common/NetStructures/Character.h:
```diff
@@ -216,19 +216,19 @@ const   QString &       getMapName() const { return m_mapName; }
 
 void                serializeStats(const Character &src, BitStream &bs, bool sendAbsolute);
-uint32_t            getLevel(Character &src) const { return src.m_level; }
+uint32_t            getLevel(Character &src) { return src.m_level; }
```
Moved all of these out of the classes. But I've created some kind of linking error and I'm getting over 600 error messages that these functions are duplicated.

nemerle commented on Projects/CoX/Servers/AdminServer/CharacterDatabase.cpp:
```diff
@@ -60,13 +64,42 @@ void CharacterDatabase::on_connected(QSqlDatabase *db)
+                "title=:title, badgetitle=:badgetitle, "
+                "specialtitle=:specialtitle, supergroup_id=:supergroup_id, "
+                "options=:options, gui=:gui "
+                "WHERE id=:id ");
```
Many of those fields should be serialized into CharacterData `blob`

nemerle commented on Projects/CoX/Servers/AdminServer/CharacterDatabase.cpp:
```diff
@@ -204,36 +246,137 @@ bool CharacterDatabase::create( uint64_t gid,uint8_t slot,Character *c )
+    ":title, :badgetitle, :specialtitle, :supergroup_id, :options, :gui"
+    */
+
+    m_prepared_char_insert.bindValue(":char_level", c->m_level);
```
```cpp
#include <iostream>
#include <cereal/archives/json.hpp>
#include <cereal/types/vector.hpp>
//......
  std::stringstream ss;
  {
    cereal::JSONOutputArchive archive( ss );
    CharacterData charData;
    archive( charData );
  }
  m_prepared_char_insert.bindValue(":char_data", QByteArray::fromStdString(charData.str()));
```

broxen replied:
I'm trying to look at costume parts for inspiration here, but I'm failing to understand this. Can you point me towards an example?

I know that I need:
- new CharacterData class containing hp, end, m_current_attribs and m_max_attribs etc.
- new `serializefromdb()` and `serializetodb()` methods for serializing these values. I see an example in `Costume.cpp` for parts.

nemerle replied:
An example from anim_serializers.cpp
```cpp
template<class Archive>
void serialize(Archive & archive, TextureAnim_Data & m)
{
    archive(cereal::make_nvp("Speed",m.speed));
    archive(cereal::make_nvp("Scale",m.stScale));
    archive(cereal::make_nvp("Type",m.scrollType));
    archive(cereal::make_nvp("flags",m.flags));
}

template
void serialize<cereal::JSONOutputArchive>(cereal::JSONOutputArchive & archive, TextureAnim_Data & m);
template
void serialize<cereal::JSONInputArchive>(cereal::JSONInputArchive & archive, TextureAnim_Data & m);
template
void serialize<cereal::BinaryInputArchive>(cereal::BinaryInputArchive & archive, TextureAnim_Data & m);
template
void serialize<cereal::BinaryOutputArchive>(cereal::BinaryOutputArchive & archive, TextureAnim_Data & m);
```

broxen commented on Projects/CoX/Servers/AdminServer/CharacterDatabase.cpp:
```diff
@@ -204,36 +246,137 @@ bool CharacterDatabase::create( uint64_t gid,uint8_t slot,Character *c )
     if(!doIt(m_prepared_char_insert))
         return false;
     int64_t char_id = m_prepared_char_insert.lastInsertId().toLongLong();
+    c->m_db_id = char_id;
```
Here I'm trying to pass the Character Table row uid to c->m_db_id. Later I try setting c->m_db_id to e->m_db_id.

broxen replied:
this now works as intended

broxen commented on Projects/CoX/Servers/AdminServer/CharacterDatabase.cpp:
```diff
@@ -204,36 +246,137 @@ bool CharacterDatabase::create( uint64_t gid,uint8_t slot,Character *c )
         return false;
     return true;
 }
+bool CharacterDatabase::update( Character *c )
```
This still doesn't run and I believe it's because c->m_db_id (:id) is never defined. It's always 0.

broxen replied:
this works now, as expected

broxen commented on Projects/CoX/Servers/MapServer/WorldSimulation.cpp:
```diff
@@ -76,4 +77,10 @@ void World::updateEntity(Entity *e, const ACE_Time_Value &dT) {
+    if(e->m_type==Entity::ENT_PLAYER)
+    {
+        CharacterDatabase *char_db = AdminServer::instance()->character_db();
+            char_db->update(&e->m_char);
```
Is this the best place to put this?

nemerle replied:
The best ? no. What we should do is a 'service thread' that periodically runs over all chars, and stores their state into db, what w should do in here, is mark the character state as 'dirty'.

But this will work for now ( add a ticket to fix/design a saner db-update code ) :)

broxen replied:
Actually this will not work for now. When more than one client is connected it crashes the server. My guess is that it tries to write to the dbase simultaneously?

broxen commented:
Current status of this PR:
- `/updatechar` works and updates character in database, but I cannot call `CharacterDatabase::update()` in WorldSimulation.cpp or the server will freeze when multiple clients are attached. :man_shrugging: 
- started fleshing out a CharacterData class based upon anim_serialization. I'm not confident that I'm heading in the right direction though. Maybe it's better to just build off of the existing Character class instead?
- `getEntity()` works but will not function from different map instances and still relies on the EntityManager object created by MapInstance (or MapClient). For now this should work fine, as we only have one instance, but in the future we'll need a better way!

TODO:
- [ ]  If I can get some help figuring out how to call `getEntity()` from the `CharacterDatabase::update()` method, I can save position information to the database. First steps towards persistence!
- [ ]  May I please have some additional guidance on the character stats serialization? Am I moving in the right direction?

Once those two items are resolved, this should be ready for commit.

broxen commented:
Identified a regression disabling entity rotation caused by converting m_orientation_pyr from float to vec3. But rotation is still choppy. Not sure how to speed up those updates.

broxen commented:
This compiles cleanly and everything seems to work. I think it's ready to commit with the following notices:
- I'm stuck on serializing to the database, I'm looking to Costume update, but missing something. So I think I've serialized everything in Character and Entity classes, but we're not currently doing anything with that JSON data, and frankly, I'm not sure how to call it elsewhere.
- I was unable to pull `m_target_idx` out of `inp_state` due to scope. InputStorage is unaware of the current entity. 
- I am unable to run `CharacterDatabase::update()` during `World::updateEntity()` because it freezes the clients with more than one character logged in.

There's still plenty to do, but I think this PR is "complete" and achieved it's original goals (and then some)
